### ISSUE TITLE
[Train] Better Hugging Face Accelerate Integration 

### ISSUE THREAD

**Issue by Yard1** [ray-project/ray#32572]
### Description

Hugging Face Accelerate is a popular library providing a higher level API over different PyTorch distributed frameworks. It already works with Ray Train, but the user is required to set a lot of configuration manually. We have several options for making user experience better.

Accelerate allows the user to create a configuration file through a CLI utility `accelerate config`, and then use that file to launch a single training worker by using `accelerate launch`, which parses the file and sets various environment variables in the process. We can mimic this behavior on our end in the Ray Train workers.

## ~1. Do not introduce any new APIs and only set the required environment variables in workers~

Already covered by https://github.com/ray-project/ray/pull/32450
This would make Accelerate (and also other libraries like DeepSpeed) work by setting the necessary PyTorch Distributed environment variables. However, it would not change the user experience that much in regards to configuring Accelerate.

```python
def training_loop(config):
    # User sets up accelerate manually, eg.
    deepspeed_plugin = DeepSpeedPlugin(
        gradient_clipping=1.0,
        zero_stage=3,
        zero3_init_flag=True,
        zero3_save_16bit_model=True,
    )
    accelerator = Accelerator(
        mixed_precision="bf16",
        deepspeed_plugin=deepspeed_plugin,
    )

    # rest of the training code

trainer = TorchTrainer(
    training_loop,
)
results = trainer.fit()
```

## ~2. Add utility functions to read and apply the config~

We can help users apply their Accelerate configs in Ray Train workers by providing two utility functions:
```python
def read_accelerate_config(path: Optional[Path] = None):
    # Reads a config yaml created with `accelerate config` as well as the
    # DeepSpeed config json if specified in the former and returns a dict
    # If path is none, use default Accelerate path
    # Called outside the training function

def apply_accelerate_config(config: dict):
    # Mimics `accelerate launch` and applies all the environment variables
    # based on the config dict (the output of `read_accelerate_config`)
    # Called inside the training function
```

This would allow the following workload:
1. User calls `accelerate config` on the driver and creates a config file as usual
2. User calls `read_accelerate_config` and obtains a config dict
3. User passes the config dict as a part of `train_loop_per_worker_config`
4. User calls `apply_accelerate_config` at the top of their training function using the config dict

```python
def training_loop(config):
    apply_accelerate_config(config["accelerate_config"])

    accelerator = Accelerator()
    # rest of the training code

# Assumes `accelerate config` was called on the driver
accelerate_config = read_accelerate_config()
trainer = TorchTrainer(
    training_loop,
    train_loop_per_worker_config={"accelerate_config": accelerate_config}
)
results = trainer.fit()
```

## 3. Add `AccelerateTrainer` (recommended)

Add a new `AccelerateTrainer` subclassing `TorchTrainer` to do the above automatically (read the Accelerate config on intialization, pass it to the training workers and apply it automatically).

This is more heavyweight but potentially provides better UX and discoverability. The downside is less flexibility.

```python
def training_loop(config):
    accelerator = Accelerator()
    # rest of the training code

# Assumes `accelerate config` was called on the driver
# AccelerateTrainer reads the config on init
trainer = AccelerateTrainer(
    training_loop,
)
results = trainer.fit()
```

## ~4. Backend~

I don't think a new  `AccelerateBackend` is a good idea here, as backends are currently an advanced API most users don't have any idea about. Still, it could be possible to have an `AccelerateBackend`&`TorchTrainer` combo work in a similar manner to `AccelerateTrainer`.

```python
def training_loop(config):
    accelerator = Accelerator()
    # rest of the training code

# Assumes `accelerate config` was called on the driver
# AccelerateConfig (containing backend) reads the config on init
trainer = TorchTrainer(
    training_loop,
    torch_config=AccelerateConfig(path=None)
)
results = trainer.fit()
```

### Use case

Better integration with the wider ML ecosystem, making it easy for users to migrate to Ray Train

**Comment by richardliaw** [comment]
Would (1) work just by itself? Don't you still need the accelerate config somehow?

**Comment by Yard1** [comment]
With (1), as a user, you'd need to either manually create the config file on all nodes in a cluster or configure the `Accelerator` in python code. As far as I understand, it should be possible to configure everything in python.

**Comment by amogkam** [comment]
I'd vote for 3 as the most seamless UX. Having an explicit trainer makes the integration much more discoverable. Also 2 seems to have too many steps right for the user? With 3 we can just cut this down to just step 1- user just needs to call `accelerate config` on the driver.

As you mentioned, 1 is always supported regardless.

**Comment by richardliaw** [comment]
I think the problem with (3) is that you still may want to use Accelerate with HuggingFace / Pytorch Lightning, and it's not clear how to use HuggingFace Trainer with AccelerateTrainer.

I think some variation of (4) actually makes the most sense. It doesn't necessarily need to be "backend", but it also doesn't hurt to start exposing that as we now see a generic need to configure the communication backend.

**Comment by amogkam** [comment]
> you still may want to use Accelerate with HuggingFace / Pytorch Lightning

Is it possible to do this? If using Huggingface or PTL, I thought it’s not possible to use accelerate since you don’t define the training loop.

**Comment by matthewdeng** [comment]
Can we clearly define what keys/information we'll be parsing from the config file output by `accelerate config`? 

For example, this config includes information about resources/hardware, such as the number of nodes and CPU/GPU info. Would we ignore these or need to use them to populate the `ScalingConfig`?

**Comment by richardliaw** [comment]
@amogkam i guess not currently but this seems to be planned https://github.com/huggingface/accelerate/issues/144

**Comment by Yard1** [comment]
@matthewdeng The following options will be ignored and replaced by Ray Train automatically:
- Number of machines
- Number of processes
- Rank of the current machine
- GPU IDs
- Number of CPU threads per process
- IP of the head process
- Port of the head process
- Type of launcher

**Comment by gjoliver** [comment]
we can also think about this problem from another angle,
which option would allow users to get their job up and running with least amount of changes to their existing workflow or code?

**Comment by Yard1** [comment]
That would probably be (3) or (4), @gjoliver

**Comment by matthewdeng** [comment]
One concern about Option 4 I have is around extensibility. Taking HuggingFace and PyTorchLightning as examples, in the current state introducing an `AccelerateConfig` could cause issues if we ever want to introduce a `HuggingFaceConfig` or `PyTorchLightningConfig`. 

Another idea I'd want to explore is to introduce something similar to a `Callback` API. This would have the API simplicity of Option 4, and with the explicitness/composability of Option 2. 

```python
class AccelerateCallback(Callback):

    def __init__(path):
        # Load config.

    def on_worker_start():
        # Set config on worker processes.
        # TODO: find a better way to define clear ordering between this and backend_config.
```

**Comment by richardliaw** [comment]
Having a bit of trouble understanding - could you help show the code that
user would write for

1. causing conflict between the AccelerateConfig vs HuggingfaceConfig?
2. What would the user write for the callback API, assuming that they dont
need to implement it themselves?

**Comment by matthewdeng** [comment]
> 1. causing conflict between the AccelerateConfig vs HuggingfaceConfig?

If `TorchConfig` does some functionality `A`, `AccelerateConfig` does functionality `A + B`, and hypothetically `HuggingfaceConfig` does functionality `A + C`, the user would not be able to specify `A + B + C`. 

Maybe HuggingFace isn't the best example here, but the question is whether we should consider Accelerate a first-class backend of its own or an integration that can be applied on top of other backends?

> 2. What would the user write for the callback API, assuming that they dont need to implement it themselves?


Something like:

```python
trainer = TorchTrainer(
    training_loop,
    callbacks = [AccelerateCallback(path=path)]
)
```

Though we'd need to also take the existing Tune callbacks into consideration when coming up with the proper design.

**Comment by richardliaw** [comment]
So basically your point is in theory these things can be additive?

What about instead of introducing a new callback system, we just make
backend_config also possible to take in a list of configs?

    TorchTrainer(backend_config=Union(List[Config], Config))

Or we can later introduce a ChainedConfig object that takes in multiple
configs?

Not saying we should - just want to know if that’s what you mean.

**Comment by matthewdeng** [comment]
Yeah, the Accelerate integration logic is additive on top of the Torch setup.

I wouldn't want to take in a list of configs (at least with how Trainers are currently implemented) because there's a tight coupling between Config <> Backend <> Trainer. Basically, the Torch Config/Backend is a primary concept, while (from my understanding) Accelerate is a secondary add-on which cannot exist on its own. 

Going off this, another idea could be to extend the `TorchConfig` API with an optional list of Callbacks. Or maybe a better name is Plugins.

```python
class TorchConfig():
    ... 
    plugins: Optional[List[TorchPlugin]] = None
```

**Comment by richardliaw** [comment]
Can we disassociate the actual implementation detail from what the user
should expect? I actually did not realize there was a difference between
config vs backend (amd also from a user perspective I’m not sure these
concepts exist in the user guide).

Do we agree that the ideal user interface is:


TorchTrainer(train_loop, SOME_PARAM=AccelerateTHING(path=…)?

And then decide on what “SOME PARAM” and “THING” should be?

The plugin/callback system seems a bit preemptive… we don’t even have a
known use case for it, so it doesn’t make that much sense to implement such
a system for n=2. I think as long as we have an acceptable path forward to
enable the functionality in the future in a backward compatible way, we can
move forward without a plugin/callback system.

**Comment by matthewdeng** [comment]
I think that's fine as long as `SOME_PARAM` is not a replacement of `torch_config`. Let's not overload this parameter please.

But yeah, in general I do agree the callbacks/plugins are a bit overengineered, so I'm okay with us making the decision to not add them right now. Anyways, based on this conversation it seems like every class that can take on a `TorchConfig` should be able to support Accelerate. What about just reversing the order?

```python

class AccelerateConfig(): # Not a BackendConfig.
    path: str

class TorchConfig(BackendConfig):
    ... 
    accelerate_config: Optional[AccelerateConfig] = None
```

**Comment by Yard1** [comment]
@matthewdeng @amogkam and I discussed this offline and we came to the conclusion that an `AccelerateTrainer` without introducing a new backend is the best option short-term, as Backends are an advanced concept that is not well surfaced. In the longer-term, we should clarify the contract between Trainers and Backends (what goes where) and look into revisiting the Backend API to ensure it is both straightforward and extensible.

For the particular usecase, we can have `HuggingFaceTrainer` inherit from `AccelerateTrainer` once Accelerate is used in Transformers.
